### PR TITLE
crimson/os/seastore/cache: fine-grained lru cache control with GC

### PR DESCRIPTION
**Pull request by aisakaki**
GC transaction is not sourced by user behaviors, so the extent read operations from GC transaction don’t satisfy the time locality principle. These extents should not be added to LRU cache.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/cache.h [hunk 1]
```diff
@@ -464,8 +464,6 @@ class Cache {
       const auto t_src = t.get_src();
       auto f = [&t, t_src, this](CachedExtent &ext) {
 	t.add_to_read_set(CachedExtentRef(&ext));
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/cache.h:466]: t_src not used, needn't captured here.

File: src/crimson/os/seastore/cache.h [hunk 2]
```diff
@@ -551,8 +549,6 @@ class Cache {
       auto t_src = t.get_src();
       auto f = [&t, t_src, this](CachedExtent &ext) {
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/cache.h:550]: same here.

File: src/crimson/os/seastore/cache.h (hunk 1 above)
Comment (aisakaki) [reply src/crimson/os/seastore/cache.h:466]: ok.

File: src/crimson/os/seastore/cache.cc [hunk 3]
```diff
@@ -696,6 +696,13 @@ void Cache::add_extent(CachedExtentRef ref)
   }
 }
 
+void Cache::add_extent_no_cache(CachedExtentRef ref)
```
Comment (athanatos) [inline src/crimson/os/seastore/cache.cc:699]: This name is confusing -- ref does get added to the cache.  Instead, let's add an optional t_src param to add_extent and update add_extent to:

```
void Cache::add_extent(
  CachedExtentRef ref,
  const Transaction::src_t* p_src=nullptr)
{
  assert(ref->is_valid());
  extents.insert(*ref);
  if (ref->is_dirty()) {
    add_to_dirty(ref);
  } else if (p_src && !is_cleaner_transaction(t.get_src()) {
    touch_extent(*ref);
  }
}
```

Comment (athanatos) [reply src/crimson/os/seastore/cache.cc:699]: Or not optional if all users have it?

Comment (aisakaki) [reply src/crimson/os/seastore/cache.cc]: ok

Comment (aisakaki) [reply src/crimson/os/seastore/cache.cc]: all users have it except in `cache::replay_delta` while there is not a transaction. I set this param as Transaction::src_t::READ in this situation, i think it might be proper?

**Comment by aisakaki** [comment]
jenkins test make check

File: src/crimson/os/seastore/cache.cc (hunk 3 above)
Comment (athanatos) [reply src/crimson/os/seastore/cache.cc]: Seems ok

Comment (liu-chunmei) [reply src/crimson/os/seastore/cache.cc]: > all users have it except in `cache::replay_delta` while there is not a transaction. I set this param as Transaction::src_t::READ in this situation, i think it might be proper?

will this be counted in Transaction READ metrics? if yes, may cause it not accurate.  if not then no problem.

Comment (aisakaki) [reply src/crimson/os/seastore/cache.cc]: > will this be counted in Transaction READ metrics? if yes, may cause it not accurate. if not then no problem.

it seems not

File: src/crimson/os/seastore/cache.h [hunk 4]
```diff
@@ -458,9 +458,8 @@ class Cache {
       SUBTRACET(seastore_cache, "{} {}~{} is absent on t, query cache ...",
                 t, T::TYPE, offset, length);
       const auto t_src = t.get_src();
-      auto f = [&t, this](CachedExtent &ext) {
+      auto f = [&t](CachedExtent &ext) {
 	t.add_to_read_set(CachedExtentRef(&ext));
-	touch_extent(ext);
```
Comment (cyx1231st) [inline src/crimson/os/seastore/cache.h:461]: if the ext is present in cache, `get_extent<T>()` won't invoke `touch_extent()`, need to fix that.

File: src/crimson/os/seastore/cache.h [hunk 5]
```diff
@@ -292,6 +292,7 @@ class Cache {
   using get_extent_ret = get_extent_ertr::future<TCachedExtentRef<T>>;
   template <typename T, typename Func, typename OnCache>
   get_extent_ret<T> get_extent(
+    const Transaction::src_t t_src, ///< [in] transaction src type
     paddr_t offset,                ///< [in] starting addr
     seastore_off_t length,          ///< [in] length
     const src_ext_t* p_metric_key, ///< [in] cache query metric key
```
Comment (cyx1231st) [inline src/crimson/os/seastore/cache.h:298]: Now I think we can reuse this parameter (probably rename to p_src_ext) because some callers do not belong to any transaction. p_src_ext will be a nullptr in this case.

File: src/crimson/os/seastore/cache.cc (hunk 3 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.cc]: Agree with chunmei that this is not accurate as it is not a READ transaction.

File: src/crimson/os/seastore/cache.h (hunk 4 above)
Comment (aisakaki) [reply src/crimson/os/seastore/cache.h:461]: oops, my mistakes. But then it seems that the GC transaction judgement should be added to `touch_extent()` instead of `add_extent()` because the directly call of `touch_extent()` is inevitable.

Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h:461]: Reasonable to me.

File: src/crimson/os/seastore/cache.h (hunk 5 above)
Comment (aisakaki) [reply src/crimson/os/seastore/cache.h]: ok

File: src/crimson/os/seastore/cache.cc [hunk 6]
```diff
@@ -1684,7 +1689,7 @@ Cache::replay_delta(
         delta.paddr,
         delta.laddr,
         delta.length,
-        nullptr,
```
Comment (cyx1231st) [inline src/crimson/os/seastore/cache.cc:1687]: nit, drop the unrelated change.

File: src/crimson/os/seastore/cache.cc [hunk 7]
```diff
@@ -1878,31 +1883,31 @@ Cache::get_extent_ertr::future<CachedExtentRef> Cache::_get_extent_by_type(
       });
     case extent_types_t::OMAP_INNER:
       return get_extent<omap_manager::OMapInnerNode>(
-          offset, length, p_metric_key, std::move(extent_init_func), std::move(on_cache)
```
Comment (cyx1231st) [inline src/crimson/os/seastore/cache.cc:1881]: nit, drop the unrelated change.

File: src/crimson/os/seastore/cache.h [hunk 8]
```diff
@@ -540,14 +541,14 @@ class Cache {
     } else {
       SUBTRACET(seastore_cache, "{} {}~{} {} is absent on t, query cache ...",
                 t, type, offset, length, laddr);
+      const auto t_src = t.get_src();
       auto f = [&t, this](CachedExtent &ext) {
 	t.add_to_read_set(CachedExtentRef(&ext));
 	touch_extent(ext);
       };
-      auto src = t.get_src();
```
Comment (cyx1231st) [inline src/crimson/os/seastore/cache.h:547]: nit, drop the unrelated change.

**Review by cyx1231st — APPROVED** [review]
LGTM

**Comment by cyx1231st** [comment]
jenkins test make check
